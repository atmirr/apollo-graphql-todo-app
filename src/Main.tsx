import React from 'react';
import { makeStyles } from '@material-ui/core/styles';
import { CssBaseline, Paper, Typography, Container } from '@material-ui/core';
import AppBar from './components/AppBar';
import AddNewTask from './components/AddNewTask';
import TasksList from './components/TasksList';

const useStyles = makeStyles(({ palette, spacing, typography }) => ({
  root: {
    display: 'flex',
    flexDirection: 'column',
    minHeight: '100vh',
    backgroundColor: palette.grey[200],
  },
  main: {
    marginTop: spacing(5),
    marginBottom: spacing(2),
    padding: spacing(3),
  },
  title: {
    fontSize: typography.pxToRem(40),
    fontWeight: typography.fontWeightLight,
    color: palette.grey[700],
  },
  footer: {
    textAlign: 'center',
    padding: spacing(3, 2),
    marginTop: 'auto',
  },
}));

function Copyright() {
  return (
    <Typography variant="subtitle2" color="textSecondary">
      {'Copyright Â©Tal&Dev '}
      {new Date().getFullYear()}
      {'.'}
    </Typography>
  );
}

function Main() {
  const classes = useStyles();
  return (
    <div className={classes.root}>
      <CssBaseline />
      <AppBar />
      <Container component="main" maxWidth="md">
        <Paper elevation={2} className={classes.main}>
          <Typography component="h1" gutterBottom className={classes.title}>
            ðŸš€ My Tasks
          </Typography>
          <AddNewTask />
          <TasksList />
        </Paper>
      </Container>
      <footer className={classes.footer}>
        <Container maxWidth="md">
          <Copyright />
        </Container>
      </footer>
    </div>
  );
}

export default Main;
