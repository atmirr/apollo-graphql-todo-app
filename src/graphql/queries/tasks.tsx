import { gql } from '@apollo/client';

const TASKS = gql`
  query {
    tasks {
      completed
      id
      parent_id
      title
    }
  }
`;

export default TASKS;
