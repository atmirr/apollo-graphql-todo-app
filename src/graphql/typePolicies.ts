import { Reference, makeVar, TypePolicies } from '@apollo/client';

export const errorVar = makeVar('');

const sortTasks = (
  tasks: Reference[] = [],
  readField: (field: string, ref: Reference) => number,
) =>
  tasks?.sort(
    (a: Reference, b: Reference) =>
      Number(readField('id', b)) - Number(readField('id', a)),
  );

export const typePolicies: TypePolicies = {
  Query: {
    fields: {
      readMainTasks(_, { readField }) {
        const tasks = readField<[]>('tasks');
        const mainTasks = tasks?.filter((taskRef) => {
          const taskParentId = readField('parent_id', taskRef);
          return taskParentId === null;
        });
        return sortTasks(mainTasks, readField);
      },
      readSubTasks(_, { variables, readField }) {
        const tasks = readField<[]>('tasks');
        const subTasks = tasks?.filter((taskRef) => {
          const taskParentId = readField('parent_id', taskRef);
          return taskParentId === variables?.parent_id;
        });
        return sortTasks(subTasks, readField);
      },
      error() {
        return errorVar();
      },
    },
  },
};
