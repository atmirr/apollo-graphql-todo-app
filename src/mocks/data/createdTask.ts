const createdTask = {
  id: 101,
  title: 'Buy pasta!',
  completed: false,
  parent_id: null,
  __typename: 'tasks',
};

export default createdTask;
