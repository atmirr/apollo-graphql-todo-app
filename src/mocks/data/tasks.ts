const tasks = [
  {
    id: 1,
    title: 'First task',
    completed: false,
    parent_id: null,
    __typename: 'tasks',
  },
  {
    id: 2,
    title: 'Second task',
    completed: true,
    parent_id: null,
    __typename: 'tasks',
  },
  {
    id: 3,
    title: 'A subtask',
    completed: false,
    parent_id: 1,
    __typename: 'tasks',
  },
];

export default tasks;
