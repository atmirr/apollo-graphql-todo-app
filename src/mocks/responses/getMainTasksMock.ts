import GET_MAIN_TASKS from 'src/graphql/queries/getMainTasks';
import tasks from '../data/tasks';

const getMainTasksMock = {
  request: {
    query: GET_MAIN_TASKS,
  },
  result: {
    data: { readMainTasks: [tasks[0], tasks[1]] },
  },
};

export default getMainTasksMock;
