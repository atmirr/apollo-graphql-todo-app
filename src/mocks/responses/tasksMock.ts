import TASKS from 'src/graphql/queries/tasks';
import tasks from '../data/tasks';

const tasksMock = {
  request: {
    query: TASKS,
  },
  result: {
    data: { tasks },
  },
};

export default tasksMock;
