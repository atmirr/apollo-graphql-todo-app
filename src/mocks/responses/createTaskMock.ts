import CREATE_TASK from 'src/graphql/mutations/createTask';
import createdTask from '../data/createdTask';

export const mutationFn = jest.fn();
const createTaskMock = {
  request: {
    query: CREATE_TASK,
    variables: { title: createdTask.title },
  },
  result: () => {
    mutationFn();
    return { data: { insert_tasks_one: createdTask } };
  },
};

export default createTaskMock;
