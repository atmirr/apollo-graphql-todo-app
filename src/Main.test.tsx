import React from 'react';
import { InMemoryCache } from '@apollo/client';
import '@testing-library/jest-dom';
import '@testing-library/jest-dom/extend-expect';
import { screen, fireEvent, findByText } from '@testing-library/react';
import { typePolicies } from 'src/graphql/typePolicies';
import tasks from 'src/mocks/data/tasks';
import createdTask from 'src/mocks/data/createdTask';
import renderWithMocksGenerator from 'src/mocks/utils/render-with-mocks';
import tasksMock from 'src/mocks/responses/tasksMock';
import getMainTasksMock from 'src/mocks/responses/getMainTasksMock';
import createTaskMock, { mutationFn } from 'src/mocks/responses/createTaskMock';
import Main from './Main';

async function waitForElements(container) {
  expect(await findByText(container, tasks[0].title)).toBeVisible();
}

let renderWithMocks;
beforeEach(() => {
  // This way we always have a fresh version
  // of the MockedProvider's cache
  // and our test units wouldn't have the side effect
  const mockedCache = new InMemoryCache({ typePolicies });
  renderWithMocks = renderWithMocksGenerator(mockedCache);
});

describe('<Main />', () => {
  it('should renders the component', async () => {
    const { container } = renderWithMocks(<Main />, [
      tasksMock,
      getMainTasksMock,
    ]);

    await waitForElements(container);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('typing in the text field and press enter should add a task into the list', async () => {
    const { container } = renderWithMocks(<Main />, [
      tasksMock,
      getMainTasksMock,
      createTaskMock,
    ]);

    await waitForElements(container);
    const inputNode = container.querySelector('input');
    fireEvent.change(inputNode, { target: { value: createdTask.title } });
    expect(inputNode.value).toBe(createdTask.title);
    fireEvent.keyDown(inputNode, { keyCode: 13 });
    expect(await findByText(container, createdTask.title)).toBeVisible();
    expect(mutationFn).toHaveBeenCalled();
  });

  it('pressing enter without adding any text should not call the mutation', async () => {
    const { container } = renderWithMocks(<Main />, [
      tasksMock,
      getMainTasksMock,
      createTaskMock,
    ]);

    await waitForElements(container);
    const inputNode = container.querySelector('input');
    expect(inputNode.value).toBe('');
    fireEvent.keyDown(inputNode, { keyCode: 13 });
    expect(screen.queryAllByText(container, createdTask.title).length).toBe(0);
    expect(mutationFn).not.toHaveBeenCalled();
  });
});
