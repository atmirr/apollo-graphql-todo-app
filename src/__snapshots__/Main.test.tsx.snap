// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Main /> should renders the component 1`] = `
<div
  class="makeStyles-root-1"
>
  <header
    class="MuiPaper-root MuiAppBar-root MuiAppBar-positionStatic MuiAppBar-colorPrimary makeStyles-appBar-5 MuiPaper-elevation4"
  >
    <div
      class="MuiToolbar-root MuiToolbar-regular MuiToolbar-gutters"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root makeStyles-icon-6"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M19 3h-4.18C14.4 1.84 13.3 1 12 1c-1.3 0-2.4.84-2.82 2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zm-7 0c.55 0 1 .45 1 1s-.45 1-1 1-1-.45-1-1 .45-1 1-1zm-2 14l-4-4 1.41-1.41L10 14.17l6.59-6.59L18 9l-8 8z"
        />
      </svg>
      <h5
        class="MuiTypography-root MuiTypography-h5"
      >
        Todo List
      </h5>
    </div>
  </header>
  <main
    class="MuiContainer-root MuiContainer-maxWidthMd"
  >
    <div
      class="MuiPaper-root makeStyles-main-2 MuiPaper-elevation2 MuiPaper-rounded"
    >
      <h1
        class="MuiTypography-root makeStyles-title-3 MuiTypography-body1 MuiTypography-gutterBottom"
      >
        🚀 My Tasks
      </h1>
      <div
        class="MuiFormControl-root MuiTextField-root makeStyles-input-7"
      >
        <div
          class="MuiInputBase-root MuiOutlinedInput-root Mui-focused Mui-focused MuiInputBase-formControl"
        >
          <input
            aria-invalid="false"
            class="MuiInputBase-input MuiOutlinedInput-input"
            placeholder="Enter a task and press Return ⏎"
            type="text"
            value=""
          />
          <fieldset
            aria-hidden="true"
            class="PrivateNotchedOutline-root-8 MuiOutlinedInput-notchedOutline"
            style="padding-left: 8px;"
          >
            <legend
              class="PrivateNotchedOutline-legend-9"
              style="width: 0.01px;"
            >
              <span>
                ​
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
      <nav
        class="MuiList-root makeStyles-root-12 MuiList-padding"
      >
        <div
          class="makeStyles-root-14"
          data-testid="task-2"
        >
          <div
            aria-disabled="false"
            class="MuiButtonBase-root MuiListItem-root makeStyles-task-15 MuiListItem-gutters MuiListItem-button"
            role="button"
            tabindex="0"
          >
            <div
              class="makeStyles-wrapper-16"
            >
              <button
                class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorPrimary"
                tabindex="0"
                type="button"
              >
                <span
                  class="MuiIconButton-label"
                >
                  <svg
                    aria-hidden="true"
                    aria-label="done"
                    class="MuiSvgIcon-root"
                    data-testid="done-2"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm-2 15l-5-5 1.41-1.41L10 14.17l7.59-7.59L19 8l-9 9z"
                    />
                  </svg>
                </span>
                <span
                  class="MuiTouchRipple-root"
                />
              </button>
              <div
                class="MuiListItemText-root makeStyles-listItemText-19 makeStyles-completedTask-18"
              >
                <span
                  class="MuiTypography-root MuiListItemText-primary makeStyles-primaryText-20 MuiTypography-body1 MuiTypography-displayBlock"
                >
                  Second task
                </span>
              </div>
            </div>
            <div
              class="makeStyles-root-22"
            >
              <button
                aria-label="add"
                class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorPrimary"
                data-testid="add-2"
                tabindex="0"
                type="button"
              >
                <span
                  class="MuiIconButton-label"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M13 7h-2v4H7v2h4v4h2v-4h4v-2h-4V7zm-1-5C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.41 0-8-3.59-8-8s3.59-8 8-8 8 3.59 8 8-3.59 8-8 8z"
                    />
                  </svg>
                </span>
                <span
                  class="MuiTouchRipple-root"
                />
              </button>
              <button
                aria-label="edit"
                class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorPrimary"
                data-testid="edit-2"
                tabindex="0"
                type="button"
              >
                <span
                  class="MuiIconButton-label"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75L3 17.25zM20.71 7.04c.39-.39.39-1.02 0-1.41l-2.34-2.34a.9959.9959 0 00-1.41 0l-1.83 1.83 3.75 3.75 1.83-1.83z"
                    />
                  </svg>
                </span>
                <span
                  class="MuiTouchRipple-root"
                />
              </button>
              <button
                aria-label="delete"
                class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorPrimary MuiIconButton-edgeEnd"
                data-testid="delete-2"
                tabindex="0"
                type="button"
              >
                <span
                  class="MuiIconButton-label"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
                    />
                  </svg>
                </span>
                <span
                  class="MuiTouchRipple-root"
                />
              </button>
            </div>
            <span
              class="MuiTouchRipple-root"
            />
          </div>
        </div>
        <div
          class="makeStyles-root-14"
          data-testid="task-1"
        >
          <div
            aria-disabled="false"
            class="MuiButtonBase-root MuiListItem-root makeStyles-task-15 MuiListItem-gutters MuiListItem-button"
            role="button"
            tabindex="0"
          >
            <div
              class="makeStyles-wrapper-16"
            >
              <button
                class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorPrimary"
                tabindex="0"
                type="button"
              >
                <span
                  class="MuiIconButton-label"
                >
                  <svg
                    aria-hidden="true"
                    aria-label="undone"
                    class="MuiSvgIcon-root"
                    data-testid="undone-1"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M16.59 7.58L10 14.17l-3.59-3.58L5 12l5 5 8-8zM12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
                    />
                  </svg>
                </span>
                <span
                  class="MuiTouchRipple-root"
                />
              </button>
              <div
                class="MuiListItemText-root makeStyles-listItemText-19"
              >
                <span
                  class="MuiTypography-root MuiListItemText-primary makeStyles-primaryText-20 MuiTypography-body1 MuiTypography-displayBlock"
                >
                  First task
                </span>
              </div>
            </div>
            <div
              class="makeStyles-root-22"
            >
              <button
                class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorPrimary"
                tabindex="0"
                type="button"
              >
                <span
                  class="MuiIconButton-label"
                >
                  <svg
                    aria-hidden="true"
                    aria-label="expand-less"
                    class="MuiSvgIcon-root"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
                    />
                  </svg>
                </span>
                <span
                  class="MuiTouchRipple-root"
                />
              </button>
              <button
                aria-label="add"
                class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorPrimary"
                data-testid="add-1"
                tabindex="0"
                type="button"
              >
                <span
                  class="MuiIconButton-label"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M13 7h-2v4H7v2h4v4h2v-4h4v-2h-4V7zm-1-5C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.41 0-8-3.59-8-8s3.59-8 8-8 8 3.59 8 8-3.59 8-8 8z"
                    />
                  </svg>
                </span>
                <span
                  class="MuiTouchRipple-root"
                />
              </button>
              <button
                aria-label="edit"
                class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorPrimary"
                data-testid="edit-1"
                tabindex="0"
                type="button"
              >
                <span
                  class="MuiIconButton-label"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75L3 17.25zM20.71 7.04c.39-.39.39-1.02 0-1.41l-2.34-2.34a.9959.9959 0 00-1.41 0l-1.83 1.83 3.75 3.75 1.83-1.83z"
                    />
                  </svg>
                </span>
                <span
                  class="MuiTouchRipple-root"
                />
              </button>
              <button
                aria-label="delete"
                class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorPrimary MuiIconButton-edgeEnd"
                data-testid="delete-1"
                tabindex="0"
                type="button"
              >
                <span
                  class="MuiIconButton-label"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
                    />
                  </svg>
                </span>
                <span
                  class="MuiTouchRipple-root"
                />
              </button>
            </div>
            <span
              class="MuiTouchRipple-root"
            />
          </div>
        </div>
      </nav>
    </div>
  </main>
  <footer
    class="makeStyles-footer-4"
  >
    <div
      class="MuiContainer-root MuiContainer-maxWidthMd"
    >
      <h6
        class="MuiTypography-root MuiTypography-subtitle2 MuiTypography-colorTextSecondary"
      >
        Copyright ©Tal&Dev 
        2021
        .
      </h6>
    </div>
  </footer>
</div>
`;
